nixos/modules/services/networking/ctrld.nix

```diff
@@ -0,0 +1,86 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.ctrld;
+  settingsFormat = pkgs.formats.toml {};
+in {
+
+  options.services.ctrld = {
+    enable = mkEnableOption "ctrld service";
+
+    package = mkPackageOption pkgs "ctrld" {};
+
+    settings = mkOption {
+      type = settingsFormat.type;
+
+      default = {};
+
+      # https://github.com/Control-D-Inc/ctrld/blob/main/README.md#default-config
+      example = {
+        listener."listener.0" = {
+          ip = "";
+          port = 0;
+          restricted = false;
+        };
+
+        network."network.0" = {
+          cidrs = [ "0.0.0.0/0" ];
+          name = "Network 0";
+        };
+
+        service = {
+          log_level = "info";
+          log_path = "";
+        };
+
+        upstream = {
+          "upstream.0" = {
+            bootstrap_ip = "76.76.2.11";
+            endpoint = "https://freedns.controld.com/p1";
+            name = "Control D - Anti-Malware";
+            timeout = 5000;
+            type = "doh";
+          };
+
+          "upstream.1" = {
+            bootstrap_ip = "76.76.2.11";
+            endpoint = "p2.freedns.controld.com";
+            name = "Control D - No Ads";
+            timeout = 3000;
+            type = "doq";
+          };
+        };
+      };
+
+      description = ''
+        Configuration for ctrld, see
+        <https://github.com/Control-D-Inc/ctrld/blob/release-branch-v${pkgs.ctrld.version}/docs/config.md>
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.nameservers = mkDefault [ "127.0.0.1" ];
```

Comment:
I'm not too sure we should be as careful as to use `mkDefault` here. I imagine a setting where the user enables this service, forgetting that they set their name server to `8.8.8.8` or whatever, and now everything doesn't work for some reason.

Reply:
Ok. Done.